presto-main/src/main/java/com/facebook/presto/operator/scalar/JsonOperators.java

```diff
@@ -13,20 +13,153 @@
 import static com.facebook.presto.metadata.OperatorType.HASH_CODE;
 import static com.facebook.presto.metadata.OperatorType.NOT_EQUAL;
+import static com.facebook.presto.spi.StandardErrorCode.INVALID_CAST_ARGUMENT;
+import static com.facebook.presto.spi.type.StandardTypes.JSON;
+import static com.fasterxml.jackson.core.JsonFactory.Feature.CANONICALIZE_FIELD_NAMES;
+import static java.lang.String.format;
 
 public final class JsonOperators
 {
+    private static final JsonFactory JSON_FACTORY = new JsonFactory().disable(CANONICALIZE_FIELD_NAMES);
+
     private JsonOperators()
     {
     }
 
+    @ScalarOperator(OperatorType.CAST)
+    @SqlType(StandardTypes.VARCHAR)
+    public static Slice castToVarchar(@SqlType(StandardTypes.JSON) Slice slice)
+    {
+        return slice;
+    }
+
+    @ScalarOperator(OperatorType.CAST)
+    @SqlType(StandardTypes.BIGINT)
+    public static long castToBigint(@SqlType(StandardTypes.JSON) Slice json)
+    {
+        try {
+            try (JsonParser parser = JSON_FACTORY.createJsonParser(json.getInput())) {
+                parser.nextToken();
+                return parser.getLongValue();
+            }
+        }
+        catch (Exception e) {
+            throw new PrestoException(INVALID_CAST_ARGUMENT, format("Cannot cast '%s' to %s", json.toStringUtf8(), StandardTypes.BIGINT));
+        }
+    }
+
+    @ScalarOperator(OperatorType.CAST)
+    @SqlType(StandardTypes.DOUBLE)
+    public static double castToDouble(@SqlType(StandardTypes.JSON) Slice json)
+    {
+        try {
+            try (JsonParser parser = JSON_FACTORY.createJsonParser(json.getInput())) {
+                parser.nextToken();
+                return parser.getDoubleValue();
+            }
+        }
+        catch (Exception e) {
+            throw new PrestoException(INVALID_CAST_ARGUMENT, format("Cannot cast '%s' to %s", json.toStringUtf8(), StandardTypes.DOUBLE));
+        }
+    }
+
+    @ScalarOperator(OperatorType.CAST)
+    @SqlType(StandardTypes.BOOLEAN)
+    public static boolean castToBoolean(@SqlType(StandardTypes.JSON) Slice json)
+    {
+        try {
+            try (JsonParser parser = JSON_FACTORY.createJsonParser(json.getInput())) {
+                parser.nextToken();
+                return parser.getBooleanValue();
+            }
+        }
+        catch (Exception e) {
+            throw new PrestoException(INVALID_CAST_ARGUMENT, format("Cannot cast '%s' to %s", json.toStringUtf8(), StandardTypes.BOOLEAN));
+        }
+    }
+
+    @ScalarOperator(OperatorType.CAST)
+    @SqlType(StandardTypes.JSON)
+    public static Slice castToJson(@SqlType(StandardTypes.VARCHAR) Slice slice) throws IOException
+    {
+        try {
+            byte[] input = slice.getBytes();
+            SliceOutput dynamicSliceOutput = new DynamicSliceOutput(input.length);
+            try (JsonGenerator jsonGenerator = JSON_FACTORY.createJsonGenerator(dynamicSliceOutput)) {
+                jsonGenerator.writeRaw(slice.toStringUtf8());
```

Comment:
The problem with the old code, I think, is it doesn't allow an arbitrary string to be cast to json while the spec allows that (please see this [doc](http://www.ecma-international.org/publications/files/ECMA-ST/ECMA-404.pdf), according to section 5, a string is a valid json value. The section 3 of the latest JSON [RFC](https://tools.ietf.org/html/rfc7159) also tells the same thing.) For example, ``cast('x' as JSON)`` will fail with the old code because of this:

```
SORTED_MAPPER.readValue(in, Object.class)
```